File: domain/src/main/java/info/firozansari/domain/interactor/GetUserDetails.java

package info.firozansari.domain.interactor;

import info.firozansari.domain.User;
import info.firozansari.domain.executor.PostExecutionThread;
import info.firozansari.domain.executor.ThreadExecutor;
import info.firozansari.domain.repository.UserRepository;
import io.reactivex.Observable;

/**
 * This class is an implementation of {@link UseCase} that represents a use case for
 * retrieving data related to an specific {@link User}.
 */
public class GetUserDetails extends UseCase<User, GetUserDetails.Params> {

  private final UserRepository userRepository;

  GetUserDetails(UserRepository userRepository, ThreadExecutor threadExecutor,
      PostExecutionThread postExecutionThread) {
    super(threadExecutor, postExecutionThread);
    this.userRepository = userRepository;
  }

  @Override Observable<User> buildUseCaseObservable(Params params) {
    return this.userRepository.user(params.userId);
  }

  public static final class Params {

    private final int userId;

    private Params(int userId) {
      this.userId = userId;
    }

    public static Params forUser(int userId) {
      return new Params(userId);
    }
  }
}

File: domain/src/main/java/info/firozansari/domain/interactor/GetUserDetails.kt
package info.firozansari.domain.interactor

import info.firozansari.domain.User
import io.reactivex.Observable

/**
 * This class is an implementation of [UseCase] that represents a use case for
 * retrieving data related to an specific [User].
 */
class GetUserDetails internal constructor(
    userRepository: UserRepository?, threadExecutor: ThreadExecutor?,
    postExecutionThread: PostExecutionThread?
) : UseCase<User?, GetUserDetails.Params?>(threadExecutor, postExecutionThread) {
    private val userRepository: UserRepository?
    public override fun buildUseCaseObservable(params: Params?): Observable<User?>? {
        return userRepository.user(params.userId)
    }

    class Params private constructor(private val userId: Int) {
        companion object {
            @JvmStatic
            fun forUser(userId: Int): Params? {
                return Params(userId)
            }
        }
    }

    init {
        this.userRepository = userRepository
    }
}